allennlp/nn/util.py

```diff
@@ -567,3 +567,42 @@ def logsumexp(tensor: torch.Tensor,
     else:
         stable_vec = tensor - max_score.unsqueeze(dim)
     return max_score + (stable_vec.exp().sum(dim, keepdim=keepdim)).log()
+
+
+def add_bos_eos(tensor, mask, bos, eos):
+    """
+    Add begin/end of sentence tokens to the batch of sentences.
+    Given a batch of sentences with size ``(batch_size, n_times)`` or
+    ``(batch_size, n_times, dim)`` this returns a tensor of shape ``(batch_size, n_times + 2)`` or
+    ``(batch_size, n_times + 2, dim)`` respectively.
+
+    Returns both the new tensor and updated mask
+
+    Parameters
+    ----------
+    tensor : A tensor of shape ``(batch_size, n_times)`` or ``(batch_size, n_times, dim)``
+    mask : A tensor of shape ``(batch_size, n_times)``
+    bos: For 2D input, a scalar with the <S> id. For 3D input, a tensor with length dim.
+    eos: For 2D input, a scalar with the </S> id. For 3D input, a tensor with length dim.
+    """
+    sequence_lengths = mask.sum(dim=1).data.numpy()
+    tensor_shape = list(tensor.data.shape)
+    new_shape = list(tensor_shape)
+    new_shape[1] = tensor_shape[1] + 2
+    ret = Variable(tensor.data.new(*new_shape).fill_(0))
+    if len(tensor_shape) == 2:
+        ret[:, 1:-1] = tensor
+        ret[:, 0] = bos
+        for i, j in enumerate(sequence_lengths):
```

Comment:
I think you can avoid looping over the tensor by using this: http://pytorch.org/docs/master/tensors.html#torch.Tensor.index_fill_

Reply:
I tried to get this to work with `index_fill_` but coundn't.  In any case, this loop isn't a bottle neck (I benchmarked it).